File: appleApp/Sources/RallyCore/UpdateChecker.swift
import Foundation

/// In-app update check against GitHub Releases. Mirrors `RallyUpdateManager.kt` state machine
/// minus the Android APK install step (macOS: download + mount, user drags to /Applications).
public struct RallyRelease: Sendable, Equatable {
    public var tag: String
    public var title: String
    public var notes: String
    public var pageUrl: String
    public var assetName: String
    public var assetUrl: String
    public var assetSize: Int64
    public init(tag: String, title: String, notes: String, pageUrl: String, assetName: String, assetUrl: String, assetSize: Int64) {
        self.tag = tag; self.title = title; self.notes = notes; self.pageUrl = pageUrl
        self.assetName = assetName; self.assetUrl = assetUrl; self.assetSize = assetSize
    }
}

private struct GHRelease: Decodable {
    var tag_name: String?
    var name: String?
    var body: String?
    var html_url: String?
    var assets: [GHAsset]?
}
private struct GHAsset: Decodable {
    var name: String?
    var browser_download_url: String?
    var size: Int64?
}

public enum UpdatePoll: Sendable, Equatable {
    case upToDate
    case available(RallyRelease)
    case failed(String)
}

public final class UpdateChecker: Sendable {
    public static let releasesURL = "https://api.github.com/repos/shivpatell25/rally-desktop/releases/latest"
    /// Release assets must come from our own repo (Android TRUSTED_RELEASE_PREFIX).
    public static let trustedAssetPrefix = "https://github.com/shivpatell25/rally-desktop/releases/download/"
    private let session: URLSession
    public init(session: URLSession = .shared) { self.session = session }

    public func check(currentVersion: String) async -> RallyRelease? {
        if case .available(let rel) = await poll(currentVersion: currentVersion) { return rel }
        return nil
    }

    /// Distinguishes up-to-date from network/parse/no-asset failures (nil-masked before).
    public func poll(currentVersion: String) async -> UpdatePoll {
        guard let url = URL(string: Self.releasesURL) else { return .failed("Bad update URL") }
        var req = URLRequest(url: url, timeoutInterval: 12)
        req.setValue("Rally/macOS", forHTTPHeaderField: "User-Agent")
        req.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        guard let (data, _) = try? await session.data(for: req),
              let gh = try? JSONDecoder().decode(GHRelease.self, from: data) else {
            return .failed("Couldn't reach the update server. Check your connection and retry.")
        }
        guard let tag = gh.tag_name, !tag.isEmpty else { return .failed("The release feed had no version.") }
        guard compareVersions(tag, currentVersion) > 0 else { return .upToDate }
        let assets = gh.assets ?? []
        // macOS disk image only — never fall back to a foreign artifact (APK/ZIP).
        guard let asset = assets.first(where: {
            let n = ($0.name ?? "").lowercased()
            return n.hasSuffix(".dmg") || (n.contains("mac") && (n.hasSuffix(".zip") || n.hasSuffix(".dmg")))
        }), let assetUrl = asset.browser_download_url, !assetUrl.isEmpty else {
            return .failed("Version \(tag) is available but has no macOS disk image yet.")
        }
        guard assetUrl.hasPrefix(Self.trustedAssetPrefix) else {
            return .failed("Version \(tag) offers an asset outside our releases — refusing it.")
        }
        return .available(RallyRelease(tag: tag, title: gh.name ?? tag, notes: gh.body ?? "",
            pageUrl: gh.html_url ?? "", assetName: asset.name ?? tag, assetUrl: assetUrl, assetSize: asset.size ?? 0))
    }

    /// Port of `compareVersions` in RallyUpdateManager.kt: core padded to 3,
    /// channels rank stable(3) > rc(2) > beta(1) > other(0) with numeric suffix,
    /// so beta installs are offered the stable release.
    public func compareVersions(_ first: String, _ second: String) -> Int {
        func parts(_ v: String) -> (core: [Int], channel: Int, channelNum: Int) {
            let clean = v.trimmingCharacters(in: .whitespaces).trimmingCharacters(in: CharacterSet(charactersIn: "vV")).lowercased()
            let coreStr: String
            var channel = 3
            var channelNum = 0
            if let dash = clean.firstIndex(of: "-") {
                coreStr = String(clean[..<dash])
                let suffix = String(clean[clean.index(after: dash)...])
                func num(_ s: String, _ prefix: String) -> Int? {
                    guard s.hasPrefix(prefix) else { return nil }
                    return Int(String(s.dropFirst(prefix.count).prefix(while: { $0.isNumber })))
                }
                if let n = num(suffix, "rc") { channel = 2; channelNum = n }
                else if let n = num(suffix, "beta") { channel = 1; channelNum = n }
                else { channel = 0; channelNum = Int(String(suffix.prefix(while: { $0.isNumber }))) ?? 0 }
            } else {
                coreStr = clean
            }
            var core = coreStr.split(separator: ".").map { Int($0.prefix(while: { $0.isNumber }).description) ?? 0 }
            while core.count < 3 { core.append(0) }
            return (core, channel, channelNum)
        }
        let a = parts(first), b = parts(second)
        for i in 0..<3 {
            if a.core[i] != b.core[i] { return a.core[i] < b.core[i] ? -1 : 1 }
        }
        if a.channel != b.channel { return a.channel < b.channel ? -1 : 1 }
        if a.channelNum != b.channelNum { return a.channelNum < b.channelNum ? -1 : 1 }
        return 0
    }
}

File: windowsApp/src/Rally.Core/UpdateChecker.cs
// In-app update check against GitHub Releases on the shared desktop train
// (shivpatell25/rally-desktop — same tags as macOS). Self-signed side-load:
// the checker reports availability and hands the installer URL to the UI;
// Windows never installs silently.
namespace Rally.Core;

public sealed record RallyRelease(string Tag, string PageUrl, string? AssetUrl, long? AssetSize, string? Notes);

public static class RallyInfo
{
    // Same train as macOS (packaging/package.sh version argument).
    public const string CurrentVersion = "0.4.0";
}

public sealed class UpdateChecker(HttpClient? http = null)
{
    public const string ReleasesUrl = "https://api.github.com/repos/shivpatell25/rally-desktop/releases/latest";
    private readonly HttpClient _http = http ?? new HttpClient() { Timeout = TimeSpan.FromSeconds(15) };

    public async Task<RallyRelease?> CheckAsync(CancellationToken ct = default)
    {
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, ReleasesUrl);
            req.Headers.UserAgent.ParseAdd("Rally/Windows");
            req.Headers.Accept.ParseAdd("application/vnd.github+json");
            using var res = await _http.SendAsync(req, ct).ConfigureAwait(false);
            if (!res.IsSuccessStatusCode) return null;
            using var doc = System.Text.Json.JsonDocument.Parse(
                await res.Content.ReadAsStringAsync(ct).ConfigureAwait(false));
            var root = doc.RootElement;
            if (root.ValueKind != System.Text.Json.JsonValueKind.Object) return null;
            if (root.TryGetProperty("draft", out var draft) && draft.ValueKind == System.Text.Json.JsonValueKind.True)
                return null;
            var tag = root.TryGetProperty("tag_name", out var t) ? t.GetString() ?? "" : "";
            if (tag.Length == 0 || CompareVersions(tag, RallyInfo.CurrentVersion) <= 0) return null;
            var page = root.TryGetProperty("html_url", out var h) ? h.GetString() : null;
            string? assetUrl = null;
            long? assetSize = null;
            if (root.TryGetProperty("assets", out var assets) && assets.ValueKind == System.Text.Json.JsonValueKind.Array)
            {
                foreach (var a in assets.EnumerateArray())
                {
                    var name = a.TryGetProperty("name", out var n) ? n.GetString() ?? "" : "";
                    var lower = name.ToLowerInvariant();
                    var isWin = lower.Contains("win") || lower.EndsWith(".msix") || lower.EndsWith(".msixbundle")
                        || lower.EndsWith(".exe") || lower.EndsWith(".zip");
                    var isApk = lower.EndsWith(".apk");
                    var isDmg = lower.EndsWith(".dmg");
                    if (!isWin || isApk || isDmg) continue;
                    assetUrl = a.TryGetProperty("browser_download_url", out var u) ? u.GetString() : null;
                    assetSize = a.TryGetProperty("size", out var s) && s.TryGetInt64(out var sz) ? sz : null;
                    break;
                }
            }
            var notes = root.TryGetProperty("body", out var b) ? b.GetString() : null;
            return new RallyRelease(tag, page ?? ReleasesUrl, assetUrl, assetSize, notes);
        }
        catch { return null; }
    }

    // 1:1 with compareVersions in RallyUpdateManager.kt: stable > rc > beta,
    // numeric suffixes, core padded to 3.
    public static int CompareVersions(string first, string second)
    {
        static (List<int> Core, int Channel, int ChannelNum) Parts(string v)
        {
            var s = v.Trim().TrimStart('v', 'V').ToLowerInvariant();
            var channel = 3;
            var channelNum = 0;
            var dash = s.IndexOf('-');
            string core = s;
            if (dash >= 0)
            {
                core = s[..dash];
                var suffix = s[(dash + 1)..];
                if (suffix.StartsWith("rc")) { channel = 2; int.TryParse(new string(suffix.Skip(2).TakeWhile(char.IsDigit).ToArray()), out channelNum); }
                else if (suffix.StartsWith("beta")) { channel = 1; int.TryParse(new string(suffix.Skip(4).TakeWhile(char.IsDigit).ToArray()), out channelNum); }
                else { channel = 0; int.TryParse(new string(suffix.TakeWhile(char.IsDigit).ToArray()), out channelNum); }
            }
            var nums = core.Split('.').Select(p => int.TryParse(new string(p.TakeWhile(char.IsDigit).ToArray()), out var n) ? n : 0).ToList();
            while (nums.Count < 3) nums.Add(0);
            return (nums, channel, channelNum);
        }
        var a = Parts(first);
        var b = Parts(second);
        for (var i = 0; i < 3; i++)
        {
            if (a.Core[i] != b.Core[i]) return a.Core[i].CompareTo(b.Core[i]);
        }
        if (a.Channel != b.Channel) return a.Channel.CompareTo(b.Channel);
        return a.ChannelNum.CompareTo(b.ChannelNum);
    }
}
